packages/core/src/external/aws/lambda-logic/bundle-to-html-bmi.ts

```diff
@@ -1598,57 +1586,134 @@ function createGastricProceduresSection(
+    : allPanels.map(({ type, panel }) => createPanelSection(type, [panel], observations)).join("");
+
+  return createSection("Laboratory", `${panelContent}`);
+}
+
+function findMetabolicPanels(
+  diagnosticReports: DiagnosticReport[],
+  panelType: string
+): DiagnosticReport[] {
+  const matchingReports = diagnosticReports.filter(report => {
+    if (report.code?.text?.toLowerCase().includes(panelType.toLowerCase())) return true;
+
+    return report.code?.coding?.some(coding =>
+      coding.display?.toLowerCase().includes(panelType.toLowerCase())
+    );
   });
 
-  const removeDuplicate = uniqWith(observationsSortedByDate, (a, b) => {
-    const aDate = dayjs(a.effectiveDateTime).format(ISO_DATE);
-    const bDate = dayjs(b.effectiveDateTime).format(ISO_DATE);
-    const aText = a.code?.text;
-    const bText = b.code?.text;
-    const aCode = a.code?.coding?.[0]?.code;
-    const bCode = b.code?.coding?.[0]?.code;
-    if (aText === undefined || bText === undefined || aCode === undefined || bCode === undefined) {
-      return false;
-    }
-    return aDate === bDate && aText === bText && aCode === bCode;
+  return matchingReports.sort((a, b) => {
+    const dateA = a.effectiveDateTime || a.effectivePeriod?.start || a.effectivePeriod?.end || "";
+    const dateB = b.effectiveDateTime || b.effectivePeriod?.start || b.effectivePeriod?.end || "";
+    return dayjs(dateB).diff(dayjs(dateA));
   });
+}
 
-  const observationTableContents =
-    removeDuplicate.length > 0
-      ? createObservationsByDate(removeDuplicate)
-      : `        <table>
-      <tbody><tr><td>No laboratory info found</td></tr></tbody>        <table>
-      `;
+function createPanelSection(
+  panelType: string,
+  panels: DiagnosticReport[],
+  allObservations: Observation[]
+): string {
+  if (panels.length === 0) return "";
+
+  return panels
+    .map(panel => {
+      const panelDate = formatDateForDisplay(
+        panel.effectiveDateTime ?? panel.effectivePeriod?.start
+      );
+      const panelObservations = getPanelObservations(panel, allObservations);
 
-  return createSection("Laboratory", observationTableContents);
+      if (panelObservations.length === 0) return "";
+
+      return `
+      <div>
+        <h4>${panelType} Results On: ${panelDate}</h4>
+        ${createObservationTable(panelObservations)}
+      </div>
+    `;
+    })
+    .join("");
 }
 
-function createObservationsByDate(observations: Observation[]): string {
+function getPanelObservations(
+  panel: DiagnosticReport,
+  allObservations: Observation[]
+): Observation[] {
+  if (!panel.result || panel.result.length === 0) return [];
+
+  return panel.result
+    .map(reference => {
+      const observationId = reference.reference?.split("/")[1];
+      return allObservations.find(obs => obs.id === observationId);
+    })
+    .filter((obs): obs is Observation => obs !== undefined);
```

Comment:
do you need this type narrowing thing here? we're just checking undefined right?

Reply:
Addressed